mivideo/mivideo_extract.py

```diff
@@ -297,31 +296,45 @@ def _makeCourseData(resultDictionaries: Sequence[Dict], categoryFilter: str) ->
         We want only those IDs that are integers, not hexadecimal or other strings.
 
         :param resultDictionaries: a sequence of KalturaMediaEntry objects converted to dictionaries
-        :param categoryFilter: the base Kaltura category, usually `Canvas_UMich`
         :return:
         """
         courseData: pd.DataFrame = pd.DataFrame.from_records(
             resultDictionaries, columns=('id', 'categories',)
-        ).rename(columns={'id': 'media_id', 'categories': 'course_id', })
+        ).rename(columns={'id': 'media_id', })
 
-        courseData = courseData.assign(
-            course_id=courseData['course_id'].str.split(',')).explode('course_id')
+        # split category CSVs to multiple new rows
+        courseData = (
+            courseData
+                .assign(categories=courseData['categories'].map(lambda c: c.split(',')))
+                .explode('categories')
+        )
 
-        courseData['in_context'] = [
-            c.endswith('>InContext') for c in courseData['course_id']]
+        # remove rows whose category isn't in "Canvas_UMich>site>channels>" hierarchy
+        courseData = courseData[courseData['categories']
+            .map(lambda c: c.startswith('Canvas_UMich>site>channels>'))]
 
-        courseData['course_id'] = courseData['course_id'].str.replace(
-            r'^' + categoryFilter + r'.*>([0-9]+)(>InContext)?$',
-            lambda m: m.groups()[0], regex=True
-        )
+        courseData['in_context'] = courseData['categories'] \
+            .map(lambda c: c.endswith('>InContext'))
+
+        # get course ID from '>'-delimited category hierarchy string
+        columnIndex = 3
+        courseData['course_id'] = \
+            courseData['categories'].map(lambda c: (c + '>' * columnIndex).split('>')[columnIndex])
```

Comment:
Are you adding `>>>` to the end of each string you are parsing to prevent index errors? Is this a standard practice? It seems maybe error-prone to me, and like maybe it would be much clearer to use [`str.count()`](https://docs.python.org/3/library/stdtypes.html#str.count) in a condition of some sort. Something like this?
```
def parseCategory(input: str) -> Union[str, None]:
   columnIndex = 3
   sepCount = input.count(`>`)
   if sepCount >= columnIndex:
      return input.split(`>`)[columnIndex)
   else:
     logger.warning(f'Unusual category input: {input}')
     return None
```

Reply:
Yes, I'm adding the extra delimiter characters to make sure the category string has at least the minimum number of delimiters to split on.  Then we'll end up with an array of at least that many elements.  In the case where we receive some Kaltura category string that's short, like "Language Resource Center" (no subcategory delimiters at all), then this code will return an empty string.  The subsequent code will filter that out.

I'm not sure whether it's a "standard" practice.  I've done this kind of thing before over the years.  It's not necessarily "**_non_**-standard".

Reply:
FWIW, I don't like logging the warning you suggested because anyone at any time can start adding categories to MiVideo media that doesn't conform to the pattern.  Then the logs will contain a lot of warnings that will worry maintainers, but they turn out to be nothing.

The important thing is that at this point in the code, a media has 𝒏 categories and at least one of them begins with `Canvas_UMich>site>channels>`.  This code is trying to weed out those categories that do not have that.

Reply:
The code I provided was meant to be an example, not necessarily an exact solution, since I don't know all the ins and outs of this data. You could put it at `DEBUG` level, though that still may be too many log messages, I have no idea.

Reply:
I'm not going to insist on you making any changes here, as I'm having trouble thinking of cases where this might actually cause problems. I still think something like what I provided is more clear than modifying the input string and assuming cases when an empty string is returned means there weren't `>` delimiters.

Reply:
Yeah, I know you're not insisting.  I appreciate that.  I also respect your suggestion.  **_Maybe_** my code is a little more like an incantation rather than simple and clear.

If this project weren't ramping down and I had other work to do in this code, then I probably would follow your advice write this in a clearer way.  But now, I don't feel like it's worthwhile to change, test, debug, commit, etc.